src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java

```diff
@@ -61,40 +62,61 @@
   private final HostInfo hostInfo;
   private Connection monitoringConn = null;
   private int connectionCheckIntervalMillis = Integer.MAX_VALUE;
+  private final AtomicLong lastContextUsedTimestamp = new AtomicLong();
+  private final long monitorDisposeTimeout;
+  private IMonitorService monitorService;
 
   public Monitor(ConnectionProvider connectionProvider, HostInfo hostInfo, PropertySet propertySet,
-      Log log) {
+      long monitorDisposeTimeout, Log log) {
     this.connectionProvider = connectionProvider;
     this.hostInfo = hostInfo;
     this.propertySet = propertySet;
     this.log = log;
+    this.monitorDisposeTimeout = monitorDisposeTimeout;
   }
 
   @Override
-  public void startMonitoring(MonitorConnectionContext context) {
+  public synchronized void startMonitoring(MonitorConnectionContext context) {
     this.connectionCheckIntervalMillis = Math.min(
         this.connectionCheckIntervalMillis,
         context.getFailureDetectionIntervalMillis());
-
-    context.setStartMonitorTime(this.getCurrentTimeMillis());
-    contexts.add(context);
+    final long currentTime = this.getCurrentTimeMillis();
+    context.setStartMonitorTime(currentTime);
+    this.lastContextUsedTimestamp.getAndSet(currentTime);
```

Comment:
NIT: don't need the "this" if your parameter name is different than the attribute.

(also, it wasn't here before? is it needed now?)